### PR TITLE
Finished lesson 07, 08, 09. fixed assignment for lesson 04

### PR DESCRIPTION
**Pull request by Luyao931213**

### CODE REVIEW

**Comment by ldconejo** [comment]
Hi Luyao,

Good job implementing the feedback for lesson 4. Please submit your assignments for lessons 7, 8 and 9 on open edX, so that I can grade them. Please open a new pull request once that's done.

Regards,

Luis

**Comment by Luyao931213** [comment]
Hi Luis,

Thank you for reminding me. Have updated on edx. Please check! 
Thanks!

Luyao



发自我的iPhone

**Comment by ldconejo** [comment]
Hi Luyao,

For lesson 8, implementation is good, make sure you run Pylint and correct any reported issues:

```
$ python -m pylint inventory.py 
************* Module inventory
inventory.py:59:0: C0304: Final newline missing (missing-final-newline)
inventory.py:49:4: E1124: Argument 'invoice_file' passed by position and keyword in function call (redundant-keyword-arg)
inventory.py:49:4: E1124: Argument 'customer_name' passed by position and keyword in function call (redundant-keyword-arg)
inventory.py:51:4: E1124: Argument 'invoice_file' passed by position and keyword in function call (redundant-keyword-arg)
inventory.py:51:4: E1124: Argument 'customer_name' passed by position and keyword in function call (redundant-keyword-arg)
inventory.py:53:4: E1124: Argument 'invoice_file' passed by position and keyword in function call (redundant-keyword-arg)
inventory.py:53:4: E1124: Argument 'customer_name' passed by position and keyword in function call (redundant-keyword-arg)

--------------------------------------------------------------------
Your code has been rated at -3.48/10 (previous run: -3.04/10, -0.43)
```

Regards,

Luis

**Comment by ldconejo** [comment]
Hi Luyao,

For lesson 9, you're not implementing your logger function as a decorator, make sure you correct that. For the PNG discovery script, it is not using recursion, which is the key requirement in the assignment.

Regards,

Luis

**Comment by Luyao931213** [comment]
Hi Luis，

I am sorry. I might be confused about the definition of recursion. But according to Wikipedia, a function being defined is applied within its own definition will be defined as a recursive function. In my function of pngdiscovery, it called itself within the definition…please let me know how to fix it.

Thanks!
Luyao


> 在 2019年10月3日，下午6:51，Luis Conejo-Alpizar <notifications@github.com> 写道：
> 
> Hi Luyao,
> 
> For lesson 9, you're not implementing your logger function as a decorator, make sure you correct that. For the PNG discovery script, it is not using recursion, which is the key requirement in the assignment.
> 
> Regards,
> 
>

File: students/Luyao_Xu/lesson09/pngdiscover.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    :param current_path:
+    :param ext:
+    :return:
+    """
+    current_path = Path(current_path)
+    result = [str(current_path.absolute()), []]
+
+    for path in current_path.iterdir():
+        if path.is_dir():
+            result.extend(list_files(path, ext))
```
Comment (ldconejo) [inline students/Luyao_Xu/lesson09/pngdiscover.py:27]: Hi Luyao, 

You're right, I missed the use of recursion here.

Regards,

Luis

**Pull request closed without merging**